src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs

```diff
@@ -123,12 +115,32 @@ internal static ProcessWaitState AddRef(int processId, bool isNewChild)
                 {
                     lock (s_processWaitStates)
                     {
+                        DateTime exitTime = default;
                         // We are referencing an existing process.
                         // This may be a child process, so we check s_childProcessWaitStates too.
-                        if (!s_childProcessWaitStates.TryGetValue(processId, out pws) &&
-                            !s_processWaitStates.TryGetValue(processId, out pws))
+                        if (s_childProcessWaitStates.TryGetValue(processId, out pws))
                         {
-                            pws = new ProcessWaitState(processId, isChild: false);
+                            // child process
+                        }
+                        else if (s_processWaitStates.TryGetValue(processId, out pws))
+                        {
+                            // This is best effort for dealing with recycled pids for non-child processes.
+                            // As long as we haven't observed process exit, it's safe to share the ProcessWaitState.
+                            // Once we've observed the exit, we'll create a new ProcessWaitState just in case
+                            // this may be a recycled pid.
+                            // If it wasn't, that ProcessWaitState will observe too that the process has exited.
+                            // We pass the ExitTime so it can be the same, but we'll clear it when we see there
+                            // is a live process with that pid.
+                            if (pws.GetExited(out _, refresh: false))
+                            {
+                                s_processWaitStates.Remove(processId);
+                                exitTime = pws.ExitTime;
+                                pws = null;
+                            }
+                        }
+                        if (pws == null)
+                        {
+                            pws = new ProcessWaitState(processId, isChild: false, exitTime);
```

Comment:
Can you help me understand this better?  It's not clear to me from this code nor from the comment why a different ProcessWaitState object helps with anything.  What's the key difference between the pws we nulled out and this one we're creating?

Reply:
The key difference is the new pws isn't in the exited state.
If we have a non-child Process with a recycled PID and we'd reused the exited pws, then the Process is unusable because HaveNonExitedId throws.